### PR TITLE
[Feature](remote)Add schema change for remote storage file

### LINKED ISSUE #13486: [Feature](remote)Add schema change for remote storage file
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Support schema change for rowset on remote storage.

### Use case

Call alter tablet column will cause an error for cold data

### Related issues

1.12

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by pengxiangyu**
# Proposed changes

Issue Number: close #13486

## Problem summary

Support schema change for rowset on remote storage.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [x] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [x] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [x] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [x] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by dataroaring** [comment]
We should not implement heavy schema change for cooldown, light schema change is done only in fe.  We can enable light schema change when creating a table with property light_schema_change = true.

**Review by yiguolei — CHANGES_REQUESTED** [review]
We already have light weight schema change(set a property during create table), not need link any more and S3(all object systems) not support link.

**Comment by pengxiangyu** [comment]
>

**Comment by pengxiangyu** [comment]
> We should not implement heavy schema change for cooldown, light schema change is done only in fe. We can enable light schema change when creating a table with property light_schema_change = true.

light_schema_change is only useful in alter table add value, but not alter table add key, which is still used in old case, so we must change the implement in BE.

you can't force user to create table with light_schema_change = true, when he want to use cold data in a old table, it doesn't work. This is used to compatible old table.

**Comment by pengxiangyu** [comment]
> We already have light weight schema change(set a property during create table), not need link any more and S3(all object systems) not support link.

you can't force user to create table with light_schema_change = true, when he want to use cold data in a old table, it doesn't work. This is used to compatible old table.

**Comment by yiguolei** [comment]
> 

I think we could add support to convert old table (not support light weight schema change) to new table (support light weight schema change) because we will set light_schema_change = true as default in one or two months. Then we could use light weight schema change to do such things.

**Comment by yiguolei** [comment]
we will remove linked schema change after 1.2 is released and it is stable.

**Comment by pengxiangyu** [comment]
> > 
> 
> 1. I think you could add support to convert old table (not support light weight schema change) to new table (support light weight schema change).  Then we could use light weight schema change to do such things.
> 2. we will set light_schema_change = true as default in one or two months. We will remove linked schema change after 1.2 is released and it is stable. Then most of link file related code will be removed.  I think you could help us add such feature then we could remove linked schema change more quickly.

light_schema_change is only useful in alter table add value, but not alter table add key, which is still used in old case, so we must change the implement in BE.

**Comment by yiguolei** [comment]
> light_schema_change is only useful in alter table add value, but not alter table add key, which is still used in old case, so we must change the implement in BE.

Do you mean add key could be resolved by using linked schema change? Could you please provide a case? 

I think if an old schema change could use linked schema change, then it could also use light weight schema change.

File: be/src/io/fs/s3_file_system.cpp [hunk 1]
```diff
@@ -226,7 +229,26 @@ Status S3FileSystem::delete_directory(const Path& path) {
...
 Status S3FileSystem::link_file(const Path& src, const Path& dest) {
-    return Status::NotSupported("not support");
+    auto client = get_client();
+    CHECK_S3_CLIENT(client);
+
+    Aws::S3::Model::CopyObjectRequest request;
+    auto src_key = get_key(src);
+    auto dest_key = get_key(dest);
+    request.WithBucket(_s3_conf.bucket)
+            .WithCopySource(_s3_conf.bucket + "/" + src_key)
```
Comment (yiguolei) [inline be/src/io/fs/s3_file_system.cpp:239]: Pls do not use copy to implement link, it is very expensive and not assure atomic and also there will be garbage if it failed. 
Should read and write a new rowset at application layer and then we could do gc.

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 39.74 seconds
 load time: 557 seconds
 storage size: 17154655370 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221026133910_clickbench_pr_34414.html

File: be/src/io/fs/s3_file_system.h [hunk 2]
```diff
@@ -75,7 +75,8 @@ class S3FileSystem final : public RemoteFileSystem {
     // Guarded by external lock.
     void set_sk(std::string sk) { _s3_conf.sk = std::move(sk); }
 
-private:
+    const S3Conf s3_conf() const { return _s3_conf; }
```
Comment (yiguolei) [inline be/src/io/fs/s3_file_system.h:78]: This method is not used.

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 3]
```diff
@@ -67,8 +67,13 @@ BetaRowsetWriter::~BetaRowsetWriter() {
...
     _context = rowset_writer_context;
     _rowset_meta.reset(new RowsetMeta);
-    if (_context.data_dir) {
+    if (_context.fs == nullptr && _context.data_dir) {
         _rowset_meta->set_fs(_context.data_dir->fs());
+    } else {
+        _rowset_meta->set_fs(_context.fs);
+    }
+    if (_context.resource_id.size() > 0) {
+        _rowset_meta->set_resource_id(_context.resource_id);
```
Comment (yiguolei) [inline be/src/olap/rowset/beta_rowset_writer.cpp:76]: why rowset meta need save resource_id? I did not find use of it.

File: be/src/olap/tablet.cpp [hunk 4]
```diff
@@ -1661,6 +1661,17 @@ Status Tablet::create_rowset_writer(const Version& version, const RowsetStatePB&
...
+    return create_rowset_writer(version, rowset_state, overlap, tablet_schema,
+                                oldest_write_timestamp, newest_write_timestamp, nullptr, "",
+                                rowset_writer);
+}
+
+Status Tablet::create_rowset_writer(const Version& version, const RowsetStatePB& rowset_state,
+                                    const SegmentsOverlapPB& overlap,
+                                    TabletSchemaSPtr tablet_schema, int64_t oldest_write_timestamp,
+                                    int64_t newest_write_timestamp, io::FileSystemPtr fs,
+                                    const io::ResourceId& resource_id,
```
Comment (yiguolei) [inline be/src/olap/tablet.cpp:1673]: could add an API resource_id() to filesystem iterface.

File: be/test/olap/rowset/segment_v2/bitmap_index_test.cpp [hunk 5]
```diff
@@ -35,6 +35,9 @@
 #include "util/file_utils.h"
 
 namespace doris {
+
+using FileSystemPtr = std::shared_ptr<io::FileSystem>;
```
Comment (yiguolei) [inline be/test/olap/rowset/segment_v2/bitmap_index_test.cpp:39]: if it is a shared ptr, then name it to  FileSystemSPtr
if it is a unique ptr, then name it to FileSystemUPtr
then we could distinguish them.

File: be/src/olap/rowset/beta_rowset_writer.cpp (hunk 3 above)
Comment (pengxiangyu) [reply be/src/olap/rowset/beta_rowset_writer.cpp:76]: it is used when calling function is_local()

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.5 seconds
 load time: 571 seconds
 storage size: 17154821261 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221027185618_clickbench_pr_34799.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 40.14 seconds
 load time: 561 seconds
 storage size: 17154644711 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221027171554_clickbench_pr_34952.html

File: be/src/olap/tablet.cpp [hunk 6]
```diff
@@ -1704,7 +1715,12 @@ void Tablet::_init_context_common_fields(RowsetWriterContext& context) {
     if (context.rowset_type == ALPHA_ROWSET) {
         context.rowset_type = StorageEngine::instance()->default_rowset_type();
     }
-    context.tablet_path = tablet_path();
+    context.rowset_dir = tablet_path();
+    if (context.fs != nullptr && context.fs->type() != io::FileSystemType::LOCAL) {
+        context.rowset_dir = BetaRowset::remote_tablet_path(tablet_id());
+    } else {
+        context.rowset_dir = tablet_path();
```
Comment (yiguolei) [inline be/src/olap/tablet.cpp:1722]: This line is useless.

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 7]
```diff
@@ -67,8 +67,13 @@ BetaRowsetWriter::~BetaRowsetWriter() {
 Status BetaRowsetWriter::init(const RowsetWriterContext& rowset_writer_context) {
     _context = rowset_writer_context;
     _rowset_meta.reset(new RowsetMeta);
-    if (_context.data_dir) {
+    if (_context.fs == nullptr && _context.data_dir) {
```
Comment (yiguolei) [inline be/src/olap/rowset/beta_rowset_writer.cpp:70]: if context.fs == nullptr and context.data_dir is also nullptr, then what happen?

File: be/src/olap/schema_change.cpp [hunk 8]
```diff
@@ -2374,6 +2374,16 @@ Status SchemaChangeHandler::_parse_request(const SchemaChangeParams& sc_params,
         *sc_directly = true;
     }
 
+    if (!(*sc_directly) && !(*sc_sorting)) {
```
Comment (yiguolei) [inline be/src/olap/schema_change.cpp:2377]: Please add a comment on line 2377, explain why we add this logic.

File: be/src/olap/rowset/beta_rowset_writer.cpp (hunk 7 above)
Comment (pengxiangyu) [reply be/src/olap/rowset/beta_rowset_writer.cpp:70]: then _rowset_meta->set_fs(nullptr), this is the same with _rowset_meta->set_fs(_context.fs);

**Review by yiguolei — DISMISSED** [review]
LGTM

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.66 seconds
 load time: 569 seconds
 storage size: 17154644880 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221028110551_clickbench_pr_35053.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.46 seconds
 load time: 571 seconds
 storage size: 17154644848 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221028112724_clickbench_pr_35071.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 39.43 seconds
 load time: 634 seconds
 storage size: 17154810725 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221028035700_clickbench_pr_35100.html

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**